DOC [PST] better integrate gallery with `pydata-sphinx-theme`

Charlie-XIAO opened the pull request:
**Please note that this PR targets the `new_web_theme` branch!**

Towards #28084. This is mainly inspired by the suggestions in [#26809 (comment)](https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1632452799).

- More detailed explanations: [#28084 (comment)](https://github.com/scikit-learn/scikit-learn/pull/28415#issuecomment-1941857261)
- Check the rendered docs [here](https://output.circle-artifacts.com/output/job/4f50d7a4-46a0-4d06-a8bc-d477f10199c5/artifacts/0/doc/auto_examples/index.html)

Here is the list of things left to be done in this PR (if any). Maintainers please feel free to modify.

- [x] Placeholder.

Charlie-XIAO committed:
[doc build] DOC [PST] better integrate gallery with pydata-sphinx-theme

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/new_web_theme' into pst-gallery

Charlie-XIAO commented:
#### Why configuring `secondary_sidebar_items` like this?

For this option when specifying a dictionary we need to use [glob-style patterns](https://www.sphinx-doc.org/en/master/usage/configuration.html#id12) and are not allowed to use multiple wild cards. Hence I traversed the directory to configure each document individually. Not sure if there is a better approach.

#### Why registering to the `"build-finished"` event?

- `"builder-inited"`: With this we need to tweak the `rst` to add some metadata, and then create components that react to this metadata and do something. This can be too complicated.
- `"html-page-context"`: First I'm not good at dealing with the [`docutils` document tree](https://tristanlatr.github.io/apidocs/docutils/docutils.nodes.html). On the other hand, `sphinx-gallery` does register some functions to the `"build-finished"` event. Relevant or not, from my perspective it is better to tweak after everything else is set.

I may not explain in detail what `sphinxext/move_gallery_links.py` is doing because I've already written comments and the code is relatively easy to understand. But feel free to leave comments if there is anything I should clarify.

#### Why modifying `README.txt`?

Essentially I just need some text there. Without the text there will be a strange vertical spacing between the first and second level headings which I'm not sure where it comes from.

Charlie-XIAO committed:
Merge remote-tracking branch 'upstream/new_web_theme' into pst-gallery

Charlie-XIAO committed:
[doc build] trigger full build

adrinjalali commented on doc/sphinxext/move_gallery_links.py:
```diff
@@ -0,0 +1,171 @@
+from pathlib import Path
```
I don't really understand what this is doing at all. And I'm not sure which problem exactly it's trying to solve. A top comment on this file would be nice.

ogrisel replied:
I think a top level comment that explains the purpose of the script is indeed needed (while the details are explained in the inline comments).

ogrisel commented on doc/conf.py:
```diff
@@ -270,7 +277,7 @@
     "footer_end": [],
     # Use :html_theme.sidebar_secondary.remove: for file-wide removal
-    "secondary_sidebar_items": ["page-toc", "sourcelink"],
+    "secondary_sidebar_items": {"**": ["page-toc", "sourcelink"]},
```
Can you please add an inline comment to explain the meaning of the `**` key in this dict?

ogrisel commented on doc/conf.py:
```diff
@@ -649,14 +658,27 @@ def reset_sklearn_config(gallery_conf, fname):
+# For each example page we use the default ["page-toc", "sourcelink"] even though we
+# only need "page-toc", because we need the secondary sidebar to exist so that
+# `sphinxext/move_gallery_links.py` can work correctly, while "page-toc" alone could
+# have been empty that causes the secondary sidebar to not exist.
```
I think this long sentence should be broken in two shorter sentences or more to be easier to understand. I think I got it but it took me several times to read it before I got it.

ogrisel commented on doc/sphinxext/move_gallery_links.py:
```diff
@@ -0,0 +1,171 @@
+
+def setup(app):
+    # Default priority is 500 which sphinx-gallery uses for its build-finished events;
+    # we need a larger (i.e. lower) priority to run after sphinx-gallery
```
```suggestion
    # we need a larger priority to run after sphinx-gallery (larger is later).
```

ogrisel commented:
I like the location of the source links in the right-hand side panel.

```css
.sidebar-secondary-item a svg {
  margin-right: 0.5em;
}
```

to get:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/bd060279-b2bc-43c3-b645-049a50a66c1e)

instead of currently:

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/877589be-3c7f-46ab-88ab-148aab58200f)

While I overall like the results, I am not a big fan of having to maintain our own HTML post-processor to move the source links at the right place.

I agree that it's a pragmatic, short-term way to move forward without having to wait for changes upstream, but in the long term, I would prefer to collaborate with sphinx-gallery maintainers to make sphinx-gallery flexible enough to have those links inserted at the right location of the pydata theme either by default or via a configuration option.

Could you please open an issue on their repo to discuss this possibility?

Charlie-XIAO committed:
[doc build] better comments

Charlie-XIAO committed:
[doc build] nits on download links appearance

Charlie-XIAO commented:
@adrinjalali @ogrisel Thanks for your review! I've added/updated some comments, please check if those are good enough.

> `.sidebar-secondary-item a svg { margin-right: 0.5em; }`

Thanks for catching this. After rechecking the other `pydata-sphinx-theme` components I found that they have a leading space in the text instead of using margin, so I updated to be consistent with their approach.

> I agree that it's a pragmatic, short-term way to move forward without having to wait for changes upstream, but in the long term, I would prefer to collaborate with sphinx-gallery maintainers to make sphinx-gallery flexible enough to have those links inserted at the right location of the pydata theme either by default or via a configuration option. Could you please open an issue on their repo to discuss this possibility?

The secondary sidebar is not a shared feature of all `sphinx` themes, and the classes and structures may differ a lot, so I may need to think more on *how* to describe the issue. Regardless I will link to this PR when opening the issue.

ogrisel commented:
Given the popularity and correlated adoptions of sphinx-gallery and the pydata sphinx theme, maybe it could make sense to add a pydata theme specific config option to sphinx-gallery?

Charlie-XIAO commented:
Yes you're right. A configuration option to hide (remove) the top note and the footer is clearly possible. As for the sidebar components, it might be possible for `pydata-sphinx-theme` to expose some API that allows people to easily create such components, or directly implement the components and somehow ship them with the extension and to allow `pydata-sphinx-theme` to discover them. I will try to formulate these thoughts upstream ASAP.

adrinjalali commented:
Awesome, so I think we can put this on hold for now and see how things go on the upstream side?

Charlie-XIAO commented:
So I've opened https://github.com/sphinx-gallery/sphinx-gallery/issues/1258, please let me know if there's anything I should add or you can directly comment in that issue.

ogrisel commented:
> Awesome, so I think we can put this on hold for now and see how things go on the upstream side?

Since this PR is targeting the `new_web_theme` branch, I am fine with merging so as to get a more representative HTML previews in other PRs targeting the same feature branch.

Then we can decide prior to merging `new_web_theme` to `main` whether or not we want to wait for the upstream discussion to settle.

adrinjalali commented:
I'd be okay with that, I like how this looks.

![image](https://github.com/scikit-learn/scikit-learn/assets/1663864/1156a043-0c42-4c24-b154-ba44012a55ec)

ogrisel commented on doc/sphinxext/move_gallery_links.py:
```diff
@@ -0,0 +1,183 @@
+
+The download links are for Python source code and Jupyter notebook respectively, and
+the badge links are for JupyterLite and Binder.
+"""
```
```suggestion

Currently this is achieved via post-processing the HTML generated by sphinx-gallery.
A possible fix upstream to avoid the need for post-processing is discussed at:
https://github.com/sphinx-gallery/sphinx-gallery/issues/1258
"""
```

Charlie-XIAO committed:
[doc build] link to upstream issue
